File: dsa_algo/patterns/merge_intervals/1_merge_intervals.py
from __future__ import print_function
class interval:
    def __init__(self, start, end):
        self.start = start
        self.end = end
    def print_interval(self):
        print("[" + str(self.start) + ", " + str(self.end) + "]", end ='')
def merge(intervals):
    if len(intervals) < 2:
        return intervals
    intervals.sort(key=lambda x: x.start)
    merged_intervals = []
    start = intervals[0].start
    end = intervals[0].end
    for i in range(1, len(intervals)):
        interval_ = intervals[i]
        if interval_.start <= end:
            end = max(interval_.end, end)
        else:
            merged_intervals.append(interval(start, end))
            start = interval_.start
            end = interval_.end
    merged_intervals.append(interval(start, end))
    return merged_intervals
def main():
    print("merged intervals: ", end='')
    for i in merge([interval(1,4), interval(2,5), interval(7,9)]):
        i.print_interval()
    print()
    
    print("merged intervals: ", end='')
    for i in merge([interval(6,7), interval(2,4), interval(5,9)]):
        i.print_interval()
    print()

    print("merged intervals: ", end='')
    for i in merge([interval(1,4), interval(2,6), interval(3,5)]):
        i.print_interval()
    print()

main()

File: dsa_algo/patterns/merge_intervals/1_merge_intervals.js
//not working. Will fix later
class interval {
    constructor(start, end)
    {
        this.start = start;
        this.end = end;
    }
    print_interval(){
        process.stdout.write(`[${this.start}, ${this.end}]`);
    }
}
const merge = (intervals) => {
    if(intervals.length < 2)
        return intervals
    intervals.sort((a,b) => a.start - b.start);
    const merged_intervals = [];
    let start = intervals[0].start;
    let end = intervals[0].end;
    for(let i = 1; i < intervals.lenght; i++)
    {
        const interval_ = intervals[i];
        if(interval_.start <= end)
            end = Math.max(interval_.end, end)
        else
        {
            merged_intervals.push(new interval(start, end));
            start = interval_.start;
            end = interval_.end;
        }
    }
    merged_intervals.push(new interval(start, end));
    return merged_intervals;
}

process.stdout.write('merged intervals: ');
let result = merge([new interval(1,4), new interval(2,5), new interval(7,9)]);
for(let i = 0; i < result.length; i++)
    result[i].print_interval();
console.log();

process.stdout.write('merged intervals: ');
result = merge([new interval(6,7), new interval(2,4), new interval(5,9)]);
for(let i = 0; i < result.length; i++)
    result[i].print_interval();
console.log();

process.stdout.write('merged intervals: ');
result = merge([new interval(1,4), new interval(2,6), new interval(3,5)]);
for(let i = 0; i < result.length; i++)
    result[i].print_interval();
console.log();
